core/src/test/java/org/apache/iceberg/variants/TestShreddedArray.java

```diff
@@ -0,0 +1,190 @@
+    for (int i = 0; i < 100_000; i++) {
+      VariantTestUtil.assertEqual(arr.get(i), elements.get(i));
+    }
+  }
+
+  @ParameterizedTest
+  @ValueSource(ints = {300, 70_000, 16_777_300})
+  public void testMultiByteOffsets(int len) {
+    // Use a string exceeding 255 bytes to test value offset sizes of 2, 3, and 4 bytes
+    String randomString = RandomUtil.generateString(len, random);
+    SerializedPrimitive bigString = VariantTestUtil.createString(randomString);
+
+    List<VariantValue> data = Lists.newArrayList();
+    data.addAll(ELEMENTS);
+    data.add(bigString);
+
+    ShreddedArray shredded = createShreddedArray(data);
+    VariantValue value = roundTripLargeBuffer(shredded);
+
+    assertThat(value.type()).isEqualTo(PhysicalType.ARRAY);
+    SerializedArray arr = (SerializedArray) value;
+    assertThat(arr.numElements()).isEqualTo(4);
+
+    assertThat(arr.get(0).type()).isEqualTo(PhysicalType.INT32);
+    assertThat(arr.get(0).asPrimitive().get()).isEqualTo(34);
+    assertThat(arr.get(1).type()).isEqualTo(PhysicalType.STRING);
+    assertThat(arr.get(1).asPrimitive().get()).isEqualTo("iceberg");
+    assertThat(arr.get(2).type()).isEqualTo(PhysicalType.DECIMAL4);
+    assertThat(arr.get(2).asPrimitive().get()).isEqualTo(new BigDecimal("12.21"));
+    assertThat(arr.get(3).type()).isEqualTo(PhysicalType.STRING);
+    assertThat(arr.get(3).asPrimitive().get()).isEqualTo(randomString);
+  }
+
+  @Test
+  public void testLargeObject() {
+    List<VariantPrimitive<String>> elements = Lists.newArrayList();
+    for (int i = 0; i < 10_000; i += 1) {
+      elements.add(Variants.of(RandomUtil.generateString(10, random)));
+    }
+
+    ShreddedArray shredded = createShreddedArray((List) elements);
+    VariantValue value = roundTripLargeBuffer(shredded);
+
+    assertThat(value.type()).isEqualTo(PhysicalType.ARRAY);
+    SerializedArray arr = (SerializedArray) value;
+    assertThat(arr.numElements()).isEqualTo(10_000);
+
+    for (VariantPrimitive<String> entry : elements) {
+      assertThat(entry.type()).isEqualTo(PhysicalType.STRING);
+      assertThat(entry.asPrimitive().get()).isEqualTo(entry.get());
+    }
+  }
+
+  private static VariantValue roundTripMinimalBuffer(ShreddedArray arr) {
+    ByteBuffer serialized = ByteBuffer.allocate(arr.sizeInBytes()).order(ByteOrder.LITTLE_ENDIAN);
+    arr.writeTo(serialized, 0);
+
+    return Variants.value(EMPTY_METADATA, serialized);
+  }
+
+  private static VariantValue roundTripLargeBuffer(ShreddedArray arr) {
+    ByteBuffer serialized =
+        ByteBuffer.allocate(1000 + arr.sizeInBytes()).order(ByteOrder.LITTLE_ENDIAN);
+    arr.writeTo(serialized, 300);
+
+    ByteBuffer slice = serialized.duplicate().order(ByteOrder.LITTLE_ENDIAN);
+    slice.position(300);
+    slice.limit(300 + arr.sizeInBytes());
+
+    return Variants.value(EMPTY_METADATA, slice);
+  }
+
+  private static ShreddedArray createShreddedArray(List<VariantValue> elements) {
```

Comment:
As with the test case above, I think using "shredded" in this name doesn't really make sense. How about `createArray`?